PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/toolkit/intellij/common/task/IntellijAzureTaskManager.java

```diff
@@ -22,63 +22,99 @@
 import com.intellij.openapi.progress.ProgressManager;
 import com.intellij.openapi.progress.Task;
 import com.intellij.openapi.project.Project;
+import com.intellij.openapi.util.Disposer;
+import com.intellij.openapi.util.registry.Registry;
 import com.microsoft.azure.toolkit.lib.common.task.AzureTask;
 import com.microsoft.azure.toolkit.lib.common.task.AzureTaskManager;
 import com.microsoft.azuretools.azurecommons.helpers.NotNull;
 
 public class IntellijAzureTaskManager extends AzureTaskManager {
 
     @Override
-    protected void doRead(final AzureTask task) {
-        ApplicationManager.getApplication().runReadAction(task.getRunnable());
+    protected void doRead(Runnable runnable, final AzureTask<?> task) {
+        ApplicationManager.getApplication().runReadAction(runnable);
     }
 
     @Override
-    protected void doWrite(final AzureTask task) {
-        ApplicationManager.getApplication().runWriteAction(task.getRunnable());
+    protected void doWrite(final Runnable runnable, final AzureTask<?> task) {
+        ApplicationManager.getApplication().runWriteAction(runnable);
     }
 
     @Override
-    protected void doRunLater(final AzureTask task) {
+    protected void doRunLater(final Runnable runnable, final AzureTask<?> task) {
         final ModalityState state = toIntellijModality(task);
-        ApplicationManager.getApplication().invokeLater(task.getRunnable(), state);
+        ApplicationManager.getApplication().invokeLater(runnable, state);
     }
 
     @Override
-    protected void doRunAndWait(final AzureTask task) {
+    protected void doRunAndWait(final Runnable runnable, final AzureTask<?> task) {
         final ModalityState state = toIntellijModality(task);
-        ApplicationManager.getApplication().invokeAndWait(task.getRunnable(), state);
+        ApplicationManager.getApplication().invokeAndWait(runnable, state);
     }
 
     @Override
-    protected void doRunInBackground(final AzureTask task) {
+    protected void doRunInBackground(final Runnable runnable, final AzureTask<?> task) {
         final Task.Backgroundable backgroundTask = new Task.Backgroundable((Project) task.getProject(), task.getTitle(), task.isCancellable()) {
             @Override
             public void run(@NotNull final ProgressIndicator progressIndicator) {
-                task.getRunnable().run();
+                task.getContext().setBackgrounded(true);
+                runnable.run();
             }
         };
         ApplicationManager.getApplication().invokeLater(() -> ProgressManager.getInstance().run(backgroundTask), ModalityState.any());
     }
 
     @Override
-    protected void doRunInModal(final AzureTask task) {
+    protected void doRunInModal(final Runnable runnable, final AzureTask<?> task) {
+        if (task.isBackgroundable()) {
+            this.doRunInBackgroundableModal(runnable, task);
+        } else {
+            this.doRunInUnBackgroundableModal(runnable, task);
+        }
```

Comment:
Unbackgroundable ?
btw, I prefer to name it as "xxxConditionalModel", just like Intellij ConditionnalModel.class.

Reply:
this method is actually implemented based on Intellij's `Task.Modal`.
with `Unbackgroudable`, I just want to emphasize that it can not be sent to background.

I actually don't understand what `ConditionalModal` is used for. and we don't use it for now.